spark-3.1-spanner-lib/src/test/java/com/google/cloud/spark/spanner/ReadIntegrationTestBase.java

```diff
@@ -58,39 +60,191 @@ public void testReadFromCompositeTable() {
     List<String> gotCs = df.select("C").as(Encoders.STRING()).collectAsList();
     List<String> expectCs = Arrays.asList("foobar", "this one");
     assertThat(gotCs).containsExactlyElementsIn(expectCs);
 
-    // 3. Validate G field boolean values.
-    List<Boolean> gotGs = df.select("G").as(Encoders.BOOLEAN()).collectAsList();
-    List<Boolean> expectGs = Arrays.asList(true, false);
-    assertThat(gotGs).containsExactlyElementsIn(expectGs);
+    // 5. Validate D field numeric values.
+    List<BigDecimal> gotDs = df.select("D").as(Encoders.DECIMAL()).collectAsList();
+    List<BigDecimal> expectDs =
+        Arrays.asList(asSparkBigDecimal("2934"), asSparkBigDecimal("93411"));
+    Collections.sort(gotDs);
+    Collections.sort(expectDs);
+    assertThat(gotDs).containsExactlyElementsIn(expectDs);
 
-    // 4. Validate E field date values.
+    // 6. Validate E field date values.
     List<Date> gotEs = df.select("E").as(Encoders.DATE()).collectAsList();
     List<Date> expectEs = Arrays.asList(Date.valueOf("2022-12-31"), Date.valueOf("2023-09-22"));
     assertThat(gotEs).containsExactlyElementsIn(expectEs);
 
-    // 5. Validate F field timestamp values.
+    // 7. Validate F field timestamp values.
     List<Timestamp> gotFs = df.select("F").as(Encoders.TIMESTAMP()).collectAsList();
     List<Timestamp> expectFs =
         Arrays.asList(
             new Timestamp(ZonedDateTime.parse("2023-08-26T12:22:05Z").toInstant().toEpochMilli()),
             new Timestamp(ZonedDateTime.parse("2023-09-22T12:22:05Z").toInstant().toEpochMilli()));
     assertThat(gotFs).containsExactlyElementsIn(expectFs);
 
-    // 6. Validate D field numeric values.
-    List<BigDecimal> gotDs = df.select("D").as(Encoders.DECIMAL()).collectAsList();
-    List<BigDecimal> expectDs =
-        Arrays.asList(asSparkBigDecimal("2934000000000"), asSparkBigDecimal("93411000000000"));
-    Collections.sort(gotDs);
-    Collections.sort(expectDs);
-    assertThat(gotDs).containsExactlyElementsIn(expectDs);
+    // 8. Validate G field boolean values.
+    List<Boolean> gotGs = df.select("G").as(Encoders.BOOLEAN()).collectAsList();
+    List<Boolean> expectGs = Arrays.asList(true, false);
+    assertThat(gotGs).containsExactlyElementsIn(expectGs);
+
+    // 9. Validate H field ARRAY<DATE> values.
+    List<Row> allHs = df.select("H").collectAsList();
+    assertEquals(allHs.size(), 2);
+    Row row1H = allHs.get(0);
+    assertEquals(row1H.size(), 1);
+    Row row2H = allHs.get(1);
+    assertEquals(row2H.size(), 1);
+    List<Date> gotHs = new ArrayList();
+    gotHs.addAll(row1H.getList(0));
+    gotHs.addAll(row2H.getList(0));
+    Collections.sort(gotHs);
+
+    List<Date> wantHs = new ArrayList();
+    Arrays.asList(
+            ZonedDateTime.parse("2023-01-02T00:00:00Z"),
+            ZonedDateTime.parse("2023-12-31T00:00:00Z"),
+            ZonedDateTime.parse("2023-09-02T00:00:00Z"),
+            ZonedDateTime.parse("2023-12-31T00:00:00Z"))
```

Comment:
We only care the year-month-date. If we get some pollution, we can verify the year, month, date separately.

Reply:
It is hell trying to use java.sql.Date and java.util.Date because they don't support specification of timezones, hence the right remedy here was to use java.time.ZonedDateTime which allows specifying ZuluTime/UTC as the timezone which helps with getting close to what Cloud Spanner returns. However, it requires that the date specify fully even the time and timezones per https://docs.oracle.com/javase/8/docs/api/java/time/ZonedDateTime.html#parse-java.lang.CharSequence-

Reply:
Okay, thanks for the clarification.